java/arcs/core/storage/CapabilitiesResolver.kt

```diff
@@ -13,25 +13,64 @@ package arcs.core.storage
 
 import arcs.core.common.ArcId
 import arcs.core.data.Capabilities
+import arcs.core.data.Schema
+import arcs.core.storage.referencemode.ReferenceModeStorageKey
 import arcs.core.util.TaggedLog
 
 /**
  * [CapabilitiesResolver] is a factory class that creates [StorageKey]s based on the registered
  * providers and given [Capabilities].
  */
 class CapabilitiesResolver(
-    val options: StorageKeyOptions,
+    val options: CapabilitiesResolverOptions,
     private val creators: StorageKeyCreatorsMap = getAllCreators()
 ) {
     private val log = TaggedLog { "CapabilitiesResolver" }
 
     /* Options used to construct [CapabilitiesResolver] */
-    data class StorageKeyOptions(val arcId: ArcId)
+    data class CapabilitiesResolverOptions(val arcId: ArcId)
+
+    /**
+     * Options passed to registered [StorageKey] constructors.
+     * @property arcId An identifier of an Arc requesting the [StorageKey]
+     * @property entitySchema A schema of an entities that will be stored
+     * @property unique A unique component of the [StorageKey]
+     * @property location A memory location of the [StorageKey]
+     */
+    interface StorageKeyOptions {
+        val arcId: ArcId
+        val entitySchema: Schema
+        val unique: String
+            get() = TODO("not implemented")
+        val location: String
+            get() = TODO("not implemented")
+    }
+
+    data class ContainerStorageKeyOptions(
+        override val arcId: ArcId,
+        override val entitySchema: Schema
+    ) : StorageKeyOptions {
+        override val unique: String
+            get() = ""
+        override val location: String
+            get() = arcId.toString()
+    }
+
+    data class BackingStorageKeyOptions(
+        override val arcId: ArcId,
+        override val entitySchema: Schema
+    ) : StorageKeyOptions {
+        override val unique: String
+            get() = entitySchema.name?.name.orEmpty()
+        override val location: String
+            get() = unique
```

Comment:
Shouldn't `unique` be the `arcId.toString()`?   If `location` is intended to be the database within which the data is stored, it would make sense to have the same location for both the container and backing keys.

Reply:
not for backing store key - it only uses entitySchema name

Reply:
What is intended to be the database on which the data is kept?  `location` as a name, would signify that it's the database where the data will be kept - to me at least

Reply:
more like `location` within the database's memory layout.. i borrowed the name from the firebase storage key, open to other ideas.